pkg/kubelet/pod_manager.go

```diff
@@ -0,0 +1,186 @@
+				podSyncTypes[pod.UID] = metrics.SyncPodCreate
+			}
+		}
+		// Actually update the pods.
+		self.pods = newPods
+		self.mirrorPods = newMirrorPods
+	case UPDATE:
+		glog.V(3).Infof("Update: Containers changed")
+
+		// Store the updated pods. Don't worry about filtering host ports since those
+		// pods will never be looked up.
+		for i := range u.Pods {
+			podSyncTypes[u.Pods[i].UID] = metrics.SyncPodUpdate
+		}
+		allPods := updatePods(u.Pods, self.pods)
+		self.pods, self.mirrorPods = filterAndCategorizePods(allPods)
+	default:
+		panic("syncLoop does not support incremental changes")
+	}
+
+	// Mark all remaining pods as sync.
+	for i := range self.pods {
+		if _, ok := podSyncTypes[self.pods[i].UID]; !ok {
+			podSyncTypes[u.Pods[i].UID] = metrics.SyncPodSync
+		}
+	}
+}
+
+func updatePods(changed []api.Pod, current []api.Pod) []api.Pod {
+	updated := []api.Pod{}
+	m := map[types.UID]*api.Pod{}
+	for i := range changed {
+		pod := &changed[i]
+		m[pod.UID] = pod
+	}
+
+	for i := range current {
+		pod := &current[i]
+		if m[pod.UID] != nil {
+			updated = append(updated, *m[pod.UID])
+			glog.V(4).Infof("pod with UID: %q has a new spec %+v", pod.UID, *m[pod.UID])
+		} else {
+			updated = append(updated, *pod)
+			glog.V(4).Infof("pod with UID: %q stay with the same spec %+v", pod.UID, *pod)
+		}
+	}
+
+	return updated
+}
+
+// GetPods returns all pods bound to the kubelet and their spec, and the mirror
+// pod map.
+func (self *podManager) GetPods() ([]api.Pod, mirrorPods) {
+	self.lock.RLock()
+	defer self.lock.RUnlock()
+	return append([]api.Pod{}, self.pods...), self.mirrorPods
+}
+
+// GetPodByName provides the first pod that matches namespace and name, as well
+// as whether the pod was found.
+func (self *podManager) GetPodByName(namespace, name string) (*api.Pod, bool) {
+	self.lock.RLock()
+	defer self.lock.RUnlock()
+	for i := range self.pods {
+		pod := self.pods[i]
+		if pod.Namespace == namespace && pod.Name == name {
+			return &pod, true
+		}
+	}
+	return nil, false
+}
+
+func (self *podManager) GetPodByFullName(podFullName string) (*api.Pod, bool) {
+	name, namespace, err := ParsePodFullName(podFullName)
+	if err != nil {
+		return nil, false
+	}
+	return self.GetPodByName(namespace, name)
+}
+
+// If the UID belongs to a mirror pod, maps it to the UID of its static pod.
+// Otherwise, return the original UID. All public-facing functions should
+// perform this translation for UIDs because user may provide a mirror pod UID,
+// which is not recognized by internal Kubelet functions.
+func (self *podManager) TranslatePodUID(uid types.UID) types.UID {
+	if uid == "" {
+		return uid
+	}
+
+	self.lock.RLock()
+	defer self.lock.RUnlock()
+	staticUID, ok := self.mirrorPods.GetStaticUID(uid)
+	if ok {
+		return staticUID
+	} else {
+		return uid
+	}
+}
+
+// Delete all orphaned mirror pods.
+func (self *podManager) DeleteOrphanedMirrorPods(mirrorPods *mirrorPods) {
+	podFullNames := mirrorPods.GetOrphanedMirrorPodNames()
```

Comment:
Yes and no. 

In kubelet, a copy of pods and mirrorPods are passed to `SyncPods()`. These copies are read-only and no one modifies it. We simply pass the read-only copy to `DeleteOrphanedMirrorPods` to avoid any lock acquiring. I added a comment here to clarify.